src/CoWSwapEthFlow.sol

```diff
@@ -0,0 +1,74 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+pragma solidity ^0.8;
+
+import "./libraries/EthFlowOrder.sol";
+import "./interfaces/ICoWSwapEthFlow.sol";
+import "./mixins/CoWSwapOnchainOrders.sol";
+
+/// @title CoW Swap ETH Flow
+/// @author CoW Swap Developers
+contract CoWSwapEthFlow is CoWSwapOnchainOrders, ICoWSwapEthFlow {
+    using EthFlowOrder for EthFlowOrder.Data;
+
+    /// @dev An order that is owned by this address is an order that has not yet been assigned.
+    address public constant NO_OWNER = address(0);
+
+    /// @dev The address of the contract representing the default native token in the current chain (e.g., WETH for
+    /// Ethereum mainnet).
+    IERC20 public immutable wrappedNativeToken;
+
+    /// @dev Each ETH flow order as described in [`EthFlowOrder.Data`] can be converted to a CoW Swap order. Distinct
+    /// CoW Swap orders have non-colliding order hashes. This mapping associates some extra data to a specific CoW Swap
+    /// order. This data is stored onchain and is used to verify the ownership and validity of an ETH flow order.
+    /// An ETH flow order can be settled onchain only if converting it to a CoW Swap order and hashing yields valid
+    /// onchain data.
+    mapping(bytes32 => EthFlowOrder.OnchainData) public orders;
+
+    /// @param _cowSwapSettlementContract The address of the CoW Swap settlement contract.
+    /// @param _wrappedNativeToken The address of the default native token in the current chain (e.g., WETH on mainnet).
+    constructor(address _cowSwapSettlementContract, IERC20 _wrappedNativeToken)
+        CoWSwapOnchainOrders(_cowSwapSettlementContract)
+    {
+        wrappedNativeToken = _wrappedNativeToken;
+    }
+
+    /// @inheritdoc ICoWSwapEthFlow
+    function createOrder(EthFlowOrder.Data calldata order)
+        external
+        payable
+        returns (bytes32 orderHash)
+    {
+        if (msg.value != order.sellAmount) {
+            revert IncorrectEthAmount();
+        }
+
+        EthFlowOrder.OnchainData memory onchainData = EthFlowOrder.OnchainData(
+            msg.sender,
```

Comment:
I plan to experiment a bit with having a `wrap` function that can be called to wrap at any time, instead of wrapping on order creation. If we have multiple ETH flow orders this could make the whole process cheaper for everyone, but I don't know how much more it costs if there is a single order or what the overhead is for cancelling an order.

Reply:
Yes, I like it. We could also only do the wrapping at settlement time by the solver. It's tricky to then calculate to redistribute the gas prices, but I think averaging it out and just putting it into the quote is fine.